src/classes/CW_AudienceCSVUploadController.cls

```diff
@@ -0,0 +1,313 @@
+	    groupsIdMap = createGroupsIdMap(campaignId);
+	    contactsIdMap = createContactsIdMap(accountNames);
+	    agsIdMap = createAgsIdMap(groupNames);
+    }
+    
+    /**/
+    
+    // CSV contains Account.Palyer_Id__c and CW_Audience_Group__r.Name
+    public static void processGroupMembers(List<List<string>> csvLines, Id campaignId,Id scheduleId)
+    {
+        List<CW_Audience_Group_Member__c> aMembers = new List<CW_Audience_Group_Member__c>();
+        Set<String> groupNames = new Set<String>();
+        Set<String> dedup = new Set<String>();  // store combination of group and playerId for de-dup
+        Schema.SObjectField agmExternalField = CW_Audience_Group_Member__c.Fields.CompositeKey__c ;
+        
+        for(List<String> flds : csvLines)
+        {
+        	if (flds.get(ConstantUtil.CSV_ACCOUNT_PLAYER_ID_COLUMN)!='')
+        	{
+	            CW_Audience_Group_Member__c aMember = new CW_Audience_Group_Member__c();
+	            // this will need to be looked up from the database
+	            aMember.Audience_Group_LU__c = groupsIdMap.get(flds.get(ConstantUtil.CSV_GROUP_NAME_COLUMN));
+	            // get contact id using the accont name from the csv file
+	            aMember.Contact__c = contactsIdMap.get(flds.get(ConstantUtil.CSV_ACCOUNT_PLAYER_ID_COLUMN));
+	            //create audience group member composite key
+	            aMember.CompositeKey__c = flds.get(ConstantUtil.CSV_ACCOUNT_PLAYER_ID_COLUMN)+flds.get(ConstantUtil.CSV_GROUP_NAME_COLUMN)+scheduleId;
+	            //create ags composite key to get the id of the ags from the agsIdMap
+	            String agsCompositeKey = flds.get(ConstantUtil.CSV_GROUP_NAME_COLUMN)+scheduleId;
+	            aMember.Audience_Group_Schedule__c = agsIdMap.get(agsCompositeKey);
+	            aMembers.add(aMember);
+        	}
+            
+        }
+        csvLines = null;
+		//upsert list of members added to the csv file
+        Database.UpsertResult[] sr = Database.upsert(aMembers,agmExternalField,false);        
+        for(Database.UpsertResult srRecord: sr)
+        {
+            if(srRecord.isSuccess())
+            {
+                System.debug('AGM Id is '+srRecord.getId());
+            }
+            else
+            {
+                for(Database.Error err : srRecord.getErrors()) 
+                {
+                    System.debug('So Error is '+err.getStatusCode() + ': ' + err.getMessage());   
+                    System.debug('Fields that affected this error: ' + err.getFields()); 
+                }
+            }
+        }
+    }
+    
+    /*** CREATE MAPS TO HELP UPLOAD GROUP MEMBERS ***/
+    
+    //create group map to get the group id based on the group name added to the csv file
+	public static Map<String,Id> createGroupsIdMap(Id campaignId)
+    {
+    	groupsIdMap = new Map<String,Id>();
+    	
+        for(CW_Audience_Group__c ag : [select id, Name from CW_Audience_Group__c where Campaign__c = :campaignId])
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.